### PR TITLE
AVM1 refactoring: moving execution from Avm1 to StackFrame

### PR DESCRIPTION
**Pull request by Dinnerbone**
Sorry for the huge PR, this took a lot of work but whilst I tried to keep each commit relatively isolated - some refactors just meant lots had to change at the exact same moment.

The crux of this is that I've moved out most things from AVM to a new StackFrame struct, which is the current activation being executed. This removes a lot unsafety (we `call current_stack_frame().unwrap()` a lot...), removes a lot of surface area for bugs and panics (the frame wasn't guaranteed to be there, or it may not be the one you're looking for), and hopefully gives a small performance boost (no need to keep fetching it over and over again). We also get to use the borrowchecker to enforce a lot of API contracts, like only executing in one frame at a time, which is a huge win.

This will also open up to solving a few issues, like being able to return from a `with {}` block, being able to give much better debugging information, and implement try/catch/finally - but I'll leave those for a future PR as I don't want to make this even bigger than it already is.

I do think it's possible to put `UpdateContext` into `StackFrame`, there shouldn't be anything preventing that, but I think this is enough of a big change for one weekend. If we were to do that, we'd only be passing around one parameter everywhere and the ergonomics could be much improved. Some day :)

### CODE REVIEW

File: core/src/avm1/property.rs [hunk 1]
```diff
@@ -52,23 +52,16 @@ impl<'gc> Property<'gc> {
...
-        this: Object<'gc>,
-        base_proto: Option<Object<'gc>>,
-        new_value: impl Into<Value<'gc>>,
-    ) -> Result<ReturnValue<'gc>, Error<'gc>> {
+    pub fn set(&mut self, new_value: impl Into<Value<'gc>>) -> Option<Executable<'gc>> {
         match self {
             Property::Virtual { set, .. } => {
                 if let Some(function) = set {
-                    function.exec(activation, context, this, base_proto, &[new_value.into()])
+                    Some(function.to_owned())
```
Comment (Herschel) [inline core/src/avm1/property.rs:42]: TODO for a later PR: This is a fairly chunky clone (`Executable` is 128+ bytes). Can we get rid of or minimize this clone while still fixing the double borrow? (One easy thing to do is to replace the block of bools in Avm1Function with an `EnumSet` save some data).

Comment (Herschel) [reply core/src/avm1/property.rs:42]: Maybe we just cram `Avm1Function` in a `Gc`.

**Comment by Herschel** [comment]
Looks good to me, thanks!

"Stack frame" and "activation" basically have the same meaning to me, so let's pick one and be consistent with terminology in variable names and comments (`stack_frame` over `activation` or vice versa).

I'm ready to merge whenever if you feel it's ready.

**Comment by Dinnerbone** [comment]
I had to separate them to begin with but merged them towards the end of the PR. At this point I can probably rename StackFrame back to Activation if you'd prefer. In code I went with `activation, context` just because `stack_frame, context` was annoying me with an underscore, no real reason other than that haha

**Comment by Dinnerbone** [comment]
@Herschel I've renamed it back to Activation, it's ready for merge :)

**Pull request merged**